cf-agent/package_module.c

```diff
@@ -0,0 +1,1611 @@
+            HasKeyDB(db_installed, package_key, strlen(package_key) + 1))
+    {
+        size_t val_size =
+                ValueSizeDB(db_installed, package_key, strlen(package_key));
+        char buff[val_size + strlen(arch) + strlen(ver) + 8];
+        
+        ReadDB(db_installed, package_key, buff, val_size);
+        xsnprintf(buff + val_size, sizeof(package_key), "V<%s>A<%s>\n", ver, arch);
+        WriteDB(db_installed, package_key, buff, strlen(buff));
+    }
+    else if (type == UPDATE_TYPE_UPDATES || type == UPDATE_TYPE_LOCAL_UPDATES)
+    {
+        char buff[strlen(arch) + strlen(ver) + 8];
+        xsnprintf(buff, sizeof(package_key), "V<%s>A<%s>\n", ver, arch);
+        WriteDB(db_installed, package_key, buff, strlen(buff));
+    }
+    else /* UPDATE_TYPE_INSTALLED */
+    {
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>V<%s>", name, ver);
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>A<%s>", name, arch);
+        WriteDB(db_installed, package_key, "1", 1);
+        xsnprintf(package_key, sizeof(package_key),
+                "N<%s>V<%s>A<%s>", name, ver, arch);
+        WriteDB(db_installed, package_key, "1", 1);
+    }
+}
+
+int UpdatePackagesDB(Rlist *data, const char *pm_name, UpdateType type)
+{
+    assert(pm_name);
+    
+    bool have_error = false;
+
+    CF_DB *db_cached;
+    dbid db_id = type == UPDATE_TYPE_INSTALLED ? dbid_packages_installed :
+                                                 dbid_packages_updates;
+    if (OpenSubDB(&db_cached, db_id, pm_name))
+    {
+        CleanDB(db_cached);
+        
+        Buffer *inventory_data = BufferNewWithCapacity(INVENTORY_LIST_BUFFER_SIZE);
+
+        char *package_data[3] = {NULL, NULL, NULL};
+
+        for (const Rlist *rp = data; rp != NULL; rp = rp->next)
+        {
+            char *line = RlistScalarValue(rp);
+
+            if (StringStartsWith(line, "Name="))
+            {
+                /* We have all the information we need from previous loop
+                 * iteration. */
+                if (package_data[0] && package_data[1] && package_data[2])
+                {
+                    WritePackageDataToDB(db_cached, package_data[0],
+                                         package_data[1], package_data[2], type);
+                    
+                    char inventory_line[strlen(package_data[0]) +
+                                        strlen(package_data[1]) +
+                                        strlen(package_data[2]) + 4];
+                    
+                    xsnprintf(inventory_line, sizeof(inventory_line),
+                              "%s,%s,%s\n", package_data[0], package_data[1],
+                              package_data[2]);
+                    BufferAppendString(inventory_data, inventory_line);
```

Comment:
Also: use `BufferAppendF(inventory_data, "%s,%s,%s\n", package_data[0], package_data[1], package_data[2]);` instead of doing your own array allocation and local xsnprintf().